spec/api-browser-window-spec.js

```diff
@@ -3222,6 +3222,29 @@ describe('BrowserWindow module', () => {
         w.loadURL('file://' + fixtures + '/api/offscreen-rendering.html')
       })
     })
+
+    describe('window.webContents.getScaleFactor()', () => {
+      it('has default scale factor', (done) => {
+        w.webContents.once('paint', function (event, rect, data) {
+          assert.equal(w.webContents.getScaleFactor(), 1.0)
+          done()
+        })
+        w.loadURL('file://' + fixtures + '/api/offscreen-rendering.html')
+      })
+    })
+
+    describe('window.webContents.setScaleFactor(scaleFactor)', () => {
+      it('sets custom scale factor', (done) => {
+        w.webContents.on('dom-ready', () => {
+          w.webContents.setScaleFactor(2.0)
+          w.webContents.once('paint', function (event, rect, data) {
+            assert.equal(w.webContents.getScaleFactor(), 2.0)
+            done()
+          })
+        })
+        w.loadURL('file://' + fixtures + '/api/offscreen-rendering.html')
+      })
+    })
```

Comment:
Please also test the actual functionality of this change, i.e. that the recorded frame has the expected size. I'd also like to see tests for the autofill popup's interactions with scale factor, since a decent amount of code here is concerned with it.

Reply:
I can, but the whole idea is that the frame size stays the same, while the window is scaled, but ofc I can test that.

Reply:
Ah, perhaps I've misunderstood the underlying intent here! You're saying the backing store remains the same size regardless of scale factor? I'm not sure I understand why that would be useful 😕

Reply:
Ahh, I see, so the frame size stays the same but the elements get bigger. I found https://cs.chromium.org/chromium/src/ui/display/display.h?l=142 which explains it pretty well—perhaps we could copy some of that descriptive text into the documentation for this API? (And perhaps call this API "SetDeviceScaleFactor" to make the terminology match up?)

I'm having a hard time googling how to read the device scale factor from JS, but if there's a way, it'd be awesome to check that in this test!

Also, is the `deviceScaleFactor` in the 'device emulation' API the same as this scale factor? https://electronjs.org/docs/api/web-contents#contentsenabledeviceemulationparameters

Reply:
`SetDeviceScaleFactor` is a great idea!

You should be able to read it from `window.devicePixelRatio`, I think you already asked about that in a comment here.

I think so, I'm gonna check if that affects OSR.

Reply:
In my experiments locally on macOS, changing the scale factor in the 'Displays' system preferences doesn't change the `devicePixelRatio` (except that in the 1:1 mode, `devicePixelRatio` reads out as 1).

<img width="824" alt="lg_ultra_hd_and_developer_tools_-_https___stackoverflow_com_questions_2242086_how-to-detect-the-screen-resolution-with-javascript" src="https://user-images.githubusercontent.com/172800/40672670-5676687c-6324-11e8-8567-f2230fff28a2.png">

<img width="814" alt="lg_ultra_hd_and_developer_tools_-_https___stackoverflow_com_questions_2242086_how-to-detect-the-screen-resolution-with-javascript" src="https://user-images.githubusercontent.com/172800/40672690-663d90f0-6324-11e8-9a38-f16d2b95f465.png">

I _think_ that's what 'device scale factor' means? Perhaps I'm still misunderstanding though!